recipes/cloudwatch_agent_install.rb

```diff
@@ -0,0 +1,86 @@
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Download the cloudwatch agent's public key
+remote_file node['cfncluster']['cloudwatch']['public_key_path'] do
+  source node['cfncluster']['cloudwatch']['public_key_url']
+  not_if { ::File.exist?(node['cfncluster']['cloudwatch']['public_key_path']) }
+end
+
+# Set URLs used to download the package and expected signature based on platform
+package_url_prefix = "https://s3.#{node['cfncluster']['cfn_region']}.amazonaws.com/amazoncloudwatch-agent-#{node['cfncluster']['cfn_region']}"
+case node['platform']
+when 'ubuntu'
+  package_url = "#{package_url_prefix}/#{node['platform']}/amd64/latest/amazon-cloudwatch-agent.deb"
+  package_path = "#{node['cfncluster']['sources_dir']}/amazon-cloudwatch-agent.rpm"
+when 'amazon'
+  # NOTE: the URL used to get amazon linux's package does not use node['platform']
+  package_url = "#{package_url_prefix}/amazon_linux/amd64/latest/amazon-cloudwatch-agent.rpm"
+  package_path = "#{node['cfncluster']['sources_dir']}/amazon-cloudwatch-agent.rpm"
+when 'centos'
+  package_url = "#{package_url_prefix}/#{node['platform']}/amd64/latest/amazon-cloudwatch-agent.rpm"
+  package_path = "#{node['cfncluster']['sources_dir']}/amazon-cloudwatch-agent.rpm"
+end
+signature_url = "#{package_url}.sig"
+signature_path = "#{package_path}.sig"
+
+# Download package and its expected signature
+remote_file signature_path do
+  source signature_url
+  not_if { ::File.exist?(signature_path) }
+end
+remote_file package_path do
+  source package_url
+  not_if { ::File.exist?(package_path) }
+end
+
+# Import cloudwatch agent's public key to the keyring
+execute "import-cloudwatch-agent-key" do
+  command "gpg --import #{node['cfncluster']['cloudwatch']['public_key_path']}"
+  user 'root'
+end
+
+# Verify that cloudwatch agent's public key has expected fingerprint
+cookbook_file 'verify_cloudwatch_agent_public_key_fingerprint.py' do
+  not_if { ::File.exist?('/usr/local/bin/verify_cloudwatch_agent_public_key_fingerprint.py') }
+  path '/usr/local/bin/verify_cloudwatch_agent_public_key_fingerprint.py'
+  user 'root'
+  group 'root'
+  mode '0755'
+end
+execute "verify-cloudwatch-agent-public-key-fingerprint" do
+  command "#{node.default['cfncluster']['cookbook_virtualenv_path']}/bin/python /usr/local/bin/verify_cloudwatch_agent_public_key_fingerprint.py"
+  user 'root'
+end
+
+# Verify that the cloudwatch agent package matches its expected signature
+execute "verify-cloudwatch-agent-rpm-signature" do
+  command "gpg --verify #{signature_path} #{package_path}"
+end
+
+# Install package.
+case node['platform']
+when 'ubuntu'
+  # Use dpkg for ubuntu because apt doesn't work
+  dpkg_package package_path do
+    source package_path
+  end
+when 'amazon', 'centos'
+  package package_path do
+    package_path
+  end
+end
```

Comment:
A single
```
  package package_path do
    source package_path
  end
```
should work

Reply:
I got failures when I did that because apparently the apt provider for the package resource does not support local installs.

Here is the output I get when trying to install:

```
Recipe: aws-parallelcluster::cloudwatch_agent_install
  * remote_file[/opt/parallelcluster/sources/amazon-cloudwatch-agent.gpg] action create (skipped due to not_if)
  * remote_file[/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb.sig] action create (skipped due to not_if)
  * remote_file[/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb] action create (skipped due to not_if)
  * execute[import-cloudwatch-agent-key] action run
    - execute gpg --import /opt/parallelcluster/sources/amazon-cloudwatch-agent.gpg
  * cookbook_file[verify_cloudwatch_agent_public_key_fingerprint.py] action create (skipped due to not_if)
  * execute[verify-cloudwatch-agent-public-key-fingerprint] action run
    - execute /root/.pyenv/versions/3.6.9/envs/cookbook_virtualenv/bin/python /usr/local/bin/verify_cloudwatch_agent_public_key_fingerprint.py
  * execute[verify-cloudwatch-agent-rpm-signature] action run
    - execute gpg --verify /opt/parallelcluster/sources/amazon-cloudwatch-agent.deb.sig /opt/parallelcluster/sources/amazon-cloudwatch-agent.deb
  * apt_package[/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb] action install
    * apt package provider cannot handle source attribute. Use dpkg provider instead
    ================================================================================
    Error executing action `install` on resource 'apt_package[/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb]'
    ================================================================================

    Chef::Exceptions::Package
    -------------------------
    apt package provider cannot handle source attribute. Use dpkg provider instead

    Resource Declaration:
    ---------------------
    # In /etc/chef/local-mode-cache/cache/cookbooks/aws-parallelcluster/recipes/cloudwatch_agent_install.rb

     87: package package_path do
     88:   source package_path
     89: end

    Compiled Resource:
    ------------------
    # Declared in /etc/chef/local-mode-cache/cache/cookbooks/aws-parallelcluster/recipes/cloudwatch_agent_install.rb:87:in `from_file'

    apt_package("/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb") do
      package_name "/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb"
      action [:install]
      default_guard_interpreter :default
      declared_type :package
      cookbook_name "aws-parallelcluster"
      recipe_name "cloudwatch_agent_install"
      source "/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb"
    end

    System Info:
    ------------
    chef_version=14.2.0
    platform=ubuntu
    platform_version=16.04
    ruby=ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]
    program_name=/usr/bin/chef-client
    executable=/opt/chef/bin/chef-client


Running handlers:
[2019-10-21T15:03:15+00:00] ERROR: Running exception handlers
Running handlers complete
[2019-10-21T15:03:15+00:00] ERROR: Exception handlers complete
Chef Client failed. 6 resources updated in 13 seconds
[2019-10-21T15:03:15+00:00] FATAL: Stacktrace dumped to /etc/chef/local-mode-cache/cache/chef-stacktrace.out
[2019-10-21T15:03:15+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2019-10-21T15:03:15+00:00] FATAL: Chef::Exceptions::Package: apt_package[/opt/parallelcluster/sources/amazon-cloudwatch-agent.deb] (aws-parallelcluster::cloudwatch_agent_install line 87) had an error: Chef::Exceptions::Package: apt package provider cannot handle source attribute. Use dpkg provider instead
```

This is a known problem based on [this issue on chef's github](https://github.com/chef/chef/issues/5760). It also looks like it's not going to get fixed anytime soon based on the fact that it's been open for 2+ years.

Conditionally using a different package resource seemed like a simpler solution than configuring a local repository for apt to use.

Reply:
After Luca and I talked briefly this morning about this, I opened [this issue](https://github.com/chef/chef/issues/9014) under chef's project. It's very similar to the one I linked to in my first response.